Must escape environment variables containing % and other special INI characters

I see a test failing with this error on Python 3.6.3:

```
$ py.test-3
================================================ test session starts =================================================
platform linux -- Python 3.6.3, pytest-3.2.3, py-1.5.2, pluggy-0.4.0                                                  
rootdir: /home/rbarlow/rpmbuild/BUILD/plaster_pastedeploy-0.4.2, inifile: setup.cfg                                   
plugins: cov-2.5.1                                         
collected 73 items                                                                                                    
                                                                                                                      
tests/test_configdict.py ..                                
tests/test_get_settings.py ....................            
tests/test_get_wsgi_app.py .....................           
tests/test_get_wsgi_app_settings.py ..F...                 
tests/test_get_wsgi_filter.py .........                    
tests/test_get_wsgi_server.py .........                    
tests/test_loader.py .                                     
tests/test_setup_logging.py .....                                                                                     
                                                                                                                      
====================================================== FAILURES ====================================================== 
__________________________________________ TestFullURI.test_foreign_config ___________________________________________

self = <tests.test_get_wsgi_app_settings.TestFullURI object at 0x7fde08d8e5c0>                                         
                                                                                                                      
    def test_foreign_config(self):                         
>       result = self.loader.get_wsgi_app_settings('test_foreign_config')                                              
                                                                                                                      
test_get_wsgi_app_settings.py:29:                                                                                     
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  
../src/plaster_pastedeploy/__init__.py:201: in get_wsgi_app_settings                                                  
    global_conf=defaults)    
/usr/lib/python3.6/site-packages/paste/deploy/loadwsgi.py:261: in appconfig                                           
    global_conf=global_conf) 
/usr/lib/python3.6/site-packages/paste/deploy/loadwsgi.py:296: in loadcontext                                         
    global_conf=global_conf)                                                                                          
/usr/lib/python3.6/site-packages/paste/deploy/loadwsgi.py:320: in _loadconfig                                         
    return loader.get_context(object_type, name, global_conf)                                                          
/usr/lib/python3.6/site-packages/paste/deploy/loadwsgi.py:454: in get_context                                         
    section)                                               
/usr/lib/python3.6/site-packages/paste/deploy/loadwsgi.py:476: in _context_from_use                                   
    object_type, name=use, global_conf=global_conf)        
/usr/lib/python3.6/site-packages/paste/deploy/loadwsgi.py:406: in get_context
    global_conf=global_conf)
/usr/lib/python3.6/site-packages/paste/deploy/loadwsgi.py:296: in loadcontext
    global_conf=global_conf)
/usr/lib/python3.6/site-packages/paste/deploy/loadwsgi.py:320: in _loadconfig
    return loader.get_context(object_type, name, global_conf)
/usr/lib/python3.6/site-packages/paste/deploy/loadwsgi.py:413: in get_context
    defaults = self.parser.defaults()
/usr/lib/python3.6/site-packages/paste/deploy/loadwsgi.py:68: in defaults
    defaults[key] = self.get('DEFAULT', key) or val
/usr/lib64/python3.6/configparser.py:800: in get
    d)
/usr/lib/python3.6/site-packages/paste/deploy/loadwsgi.py:95: in before_get
    value, defaults)
/usr/lib64/python3.6/configparser.py:394: in before_get
    self._interpolate_some(parser, option, L, value, section, defaults, 1)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <configparser.BasicInterpolation object at 0x7fde09b2c908>
parser = <paste.deploy.loadwsgi.NicerConfigParser object at 0x7fde08dbd0f0>, option = 'ENV_TMUX_PANE', accum = []
rest = '%4', section = 'DEFAULT'
map = ChainMap({}, {}, OrderedDict([('here', '/home/rbarlow/rpmbuild/BUILD/plaster_pastedeploy-0.4.2/tests/sample_configs'),...), ('basepath', '/home/rbarlow/rpmbuild/BUILD/plaster_pastedeploy-0.4.2/tests/sample_configs'), ('glob', 'overr
ide')]))
depth = 1

    def _interpolate_some(self, parser, option, accum, rest, section, map,
                          depth):
        rawval = parser.get(section, option, raw=True, fallback=rest)
        if depth > MAX_INTERPOLATION_DEPTH:
            raise InterpolationDepthError(option, section, rawval)
        while rest:
            p = rest.find("%")
            if p < 0:
                accum.append(rest)
                return
            if p > 0:
                accum.append(rest[:p])
                rest = rest[p:]
            # p is no longer used
            c = rest[1:2]
            if c == "%":
                accum.append("%")
                rest = rest[2:]
            elif c == "(":
                m = self._KEYCRE.match(rest)
                if m is None:
                    raise InterpolationSyntaxError(option, section,
                        "bad interpolation variable reference %r" % rest)
                var = parser.optionxform(m.group(1))
                rest = rest[m.end():]
                try:
                    v = map[var]
                except KeyError:
                    raise InterpolationMissingOptionError(
                        option, section, rawval, var) from None
                if "%" in v:
                    self._interpolate_some(parser, option, accum, v,
                                           section, map, depth + 1)
                else:
                    accum.append(v)
            else:
                raise InterpolationSyntaxError(
                    option, section,
                    "'%%' must be followed by '%%' or '(', "
>                   "found: %r" % (rest,))
E               configparser.InterpolationSyntaxError: Error in file /home/rbarlow/rpmbuild/BUILD/plaster_pastedeploy-0.4.2/tests/sample_configs/test_config_included.ini: '%' must be followed by '%' or '(', found: '%4'

/usr/lib64/python3.6/configparser.py:444: InterpolationSyntaxError
======================================== 1 failed, 72 passed in 2.38 seconds =========================================
```

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -1,3 +1,11 @@
+unreleased
+==========
+
+- Removed environment variable support entirely for now. The feature requires
+  bugfixes upstream in PasteDeploy which have not been done yet and this was
+  breaking people's environments so it is gone for now.
+  See https://github.com/Pylons/plaster_pastedeploy/pull/15
+
 0.4.2 (2017-11-20)
 ==================
 

--- a/README.rst
+++ b/README.rst
@@ -67,29 +67,5 @@ Some examples are below:
 
 - ``egg:MyApp?debug=false#foo``
 
-Environment Variables
----------------------
-
-**This feature is not supported on Python 2.**
-
-This binding extends ``pastedeploy`` to inject environ variables as defaults
-which are available to use in the INI file. Each ``os.environ`` key is prefixed
-with ``ENV_`` and added as a default.
-
-For example:
-
-.. code-block:: ini
-
-   [app:main]
-   debug = %(ENV_APP_DEBUG)s
-
-The only thing to be aware of here is that there is no fallback. The INI file
-will fail to parse if the environment variable is not set. The app may be run
-like so:
-
-.. code-block:: bash
-
-   $ APP_DEBUG=true env/bin/pserve development.ini
-
 .. _PasteDeploy: https://pastedeploy.readthedocs.io/en/latest/
 .. _plaster: https://docs.pylonsproject.org/projects/plaster/en/latest/

--- a/src/plaster_pastedeploy/__init__.py
+++ b/src/plaster_pastedeploy/__init__.py
@@ -227,15 +227,6 @@ class Loader(IWSGIProtocol, ILoader):
             result['here'] = os.getcwd()
         else:
             result['here'] = os.path.dirname(self.filepath)
-        if not PY2:
-            # Only inject environment variables on py3+ where escaping is
-            # supported. On py2 any environment var with contents of the
-            # format %(foo)s would break the parser. Unfortunately, this is
-            # risky enough to simply not support it.
-            result.update({
-                'ENV_' + k: v.replace('%', '%%')
-                for k, v in os.environ.items()
-            })
         result.update(self.uri.options)
         if defaults:
             result.update(defaults)

--- a/tests/sample_configs/test_settings.ini
+++ b/tests/sample_configs/test_settings.ini
@@ -12,7 +12,3 @@ set default_b = override_b
 a = a_val
 b = b_val
 c = %(default_c)s
-
-[section3]
-foo = %(ENV_PLASTER_FOO)s
-bar = %(ENV_PLASTER_BAR)s

--- a/tests/test_get_settings.py
+++ b/tests/test_get_settings.py
@@ -17,7 +17,7 @@ class TestSimpleUri(object):
 
     def test_sections(self):
         result = self.loader.get_sections()
-        assert set(result) == {'section1', 'section2', 'section3'}
+        assert set(result) == {'section1', 'section2'}
 
     def test_missing_section(self):
         result = self.loader.get_settings('missing', {'a': 'b'})
@@ -57,18 +57,6 @@ class TestSimpleUri(object):
         assert result.global_conf['default_a'] == 'default_a'
         assert result.global_conf['default_b'] == 'default_b'
         assert result.global_conf['default_c'] == 'default_c'
-
-    @pytest.mark.skipif('sys.version_info[0] == 2')
-    def test_environ_passed_and_escaped(self, monkeypatch):
-        monkeypatch.setenv('PLASTER_FOO', '%(foo)s')
-        monkeypatch.setenv('PLASTER_BAR', '%bar')
-        result = self.loader.get_settings('section3')
-
-        assert result['foo'] == '%(foo)s'
-        assert result.global_conf['ENV_PLASTER_FOO'] == '%(foo)s'
-
-        assert result['bar'] == '%bar'
-        assert result.global_conf['ENV_PLASTER_BAR'] == '%bar'
 
 class TestSectionedURI(TestSimpleUri):
     config_uri = test_settings_path + '#section1'

